---
title: "Ultra-long reads coverage of IGL locus"
author: "wujiaqi"
date: "2022-12-26"
output: html_document
---

```{r}
suppressMessages(library(trackViewer)) 
suppressMessages(library(data.table))
suppressMessages(library(RColorBrewer))
suppressMessages(library(Gviz))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg19))
```


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/home/wujiaqi/data/cow/")
```

## Figure S9A - NCBA 1.0 IGL gene loci (with gene name)

```{r}
IGL_region_2 <- as.data.frame(fread("./newData_20220115/IGL/IGL_mapResults-20220130_cleanup.csv"))
IGL_region_2$location <- IGL_region_2$start+round(0.5*IGL_region_2$length)

## 1、IGL allel graph
IGL_gr <- GRanges("IGL", IRanges(IGL_region_2$location, width = 1, names = IGL_region_2$annotation))
features_IGL <- GRanges("IGL", IRanges(start = 0, end = 72700000-72000000))

features_IGL$fill <- c("#CE896A")
features_IGL$color <- c("#CE896A")
features_IGL$height <- c(0.04)

IGL_gr$SNPsideID <- "top"
features_IGL_mul <- GRanges("IGL", IRanges(start = c(0,IGL_region_2$start),
                                           end = c(72700000-72000000,IGL_region_2$start+IGL_region_2$length)))
table(IGL_region_2$functionality)
IGL_region_2$color <- NA
IGL_region_2$color[IGL_region_2$functionality %in% "F"] <- "#c93f55"
IGL_region_2$color[IGL_region_2$functionality %in% "ORF"] <- "#ffcd12"
IGL_region_2$color[IGL_region_2$functionality %in% "P"] <- "#208cc0"
IGL_region_2$color[IGL_region_2$functionality %in% "" | IGL_region_2$functionality %in% NA] <- "#828282"

IGL_allel_num <- nrow(IGL_region_2)
# features_IGL_mul$fill <- rep(c("#44435F","#CE896A"), c(IGL_allel_num,1))
# features_IGL_mul$color <- rep(c("#44435F","#CE896A"), c(IGL_allel_num,1))
features_IGL_mul$fill <- c("#44435F",IGL_region_2$color)
features_IGL_mul$color <- c("#44435F",IGL_region_2$color)
features_IGL_mul$height <- rep(0.04, length(features_IGL_mul))
# features_IGL_mul$height[4:6] <- list(unit(1/8, "inches"),
#                                  unit(0.5, "lines"),
#                                  unit(.2, "char"))
features_IGL_mul$featureLayerID <-
  rep(c("bottom.contig", "top.IGL.allel"), c(1,IGL_allel_num))
names(features_IGL_mul) <- rep(c("contig","IGL allel"),c(1,IGL_allel_num))

IGL_gr$SNPsideID <- "top"
gr <- GRanges("IGL", IRanges(10000,240000))
xaxis_tmp <- c(10000,60000,110000,160000,210000,240000)
names(xaxis_tmp) <- c("72.01Mb","72.06Mb","72.11Mb","72.16Mb","72.21Mb","72.24Mb")
x <- lolliplot(IGL_gr, features_IGL_mul, ranges = gr, cex = 0.6, yaxis = F, ylab = T, xaxis = xaxis_tmp)


gr <- GRanges("IGL", IRanges(240000,470000))
xaxis_tmp <- c(240000,260000,310000,360000,410000,460000,470000)
names(xaxis_tmp) <- c("72.24Mb","72.26Mb","72.31Mb","72.36Mb","72.41Mb","72.46Mb","72.47Mb")
x <- lolliplot(IGL_gr, features_IGL_mul, ranges = gr, cex = 0.6, yaxis = F, ylab = T, xaxis = xaxis_tmp)


gr <- GRanges("IGL", IRanges(470000,700000))
xaxis_tmp <- c(470000,510000,560000,610000,660000,700000)
names(xaxis_tmp) <- c("72.47Mb","72.51Mb","72.56Mb","72.61Mb","72.66Mb","72.70Mb")
x <- lolliplot(IGL_gr, features_IGL_mul, ranges = gr, cex = 0.6, yaxis = F, ylab = T, xaxis = xaxis_tmp)

```


## Figure S9B - NCBA1.0 IGL gene loci (chr and contig info)

```{r}
IGL_region_2 <- as.data.frame(fread("./newData_20220115/IGL/IGL_mapResults-20220130_cleanup.csv"))
IGL_region_2$location <- 72000000+IGL_region_2$start+round(0.5*IGL_region_2$length)

chr_ctg_loci <- fread("./newData_20220115/chrContigloci.csv")


IGL_gr <- GRanges("chr17", IRanges(IGL_region_2$location, width = 1, names = IGL_region_2$annotation))
IGL_gr$score <- runif(length(IGL_gr))*10

IGL_gr$SNPsideID <- "top"
features_IGL <- GRanges("chr17", IRanges(start = c(0,0,72000000+IGL_region_2$start), end = c(74718404,74718404,72000000+IGL_region_2$start+IGL_region_2$length)))

features_IGL$fill <- rep(c("#647C9C","#CE896A","#101010"), c(1,1,nrow(IGL_region_2)))
features_IGL$color <- rep(c("#647C9C","#CE896A","#101010"), c(1,1,nrow(IGL_region_2)))
features_IGL$height <- c(0.04)
features_IGL$featureLayerID <-
  rep(c("bottom.chr", "median.contig", "top.IGL.allel"), c(1, 1,nrow(IGL_region_2)))
names(features_IGL) <- rep(c("chr17", "ctg000010", "IGL allel"), c(1, 1,nrow(IGL_region_2)))
x <- lolliplot(IGL_gr, features_IGL, cex = 0.6, yaxis = F, ylab = F)


IGL_gr <- GRanges("chr17", IRanges(-1, width = 1))
IGL_gr$score <- runif(length(IGL_gr))*10

IGL_gr$SNPsideID <- "bottom"
features_IGL <- GRanges("chr17", IRanges(start = c(72000000+IGL_region_2$start,0,0), end = c(72000000+IGL_region_2$start+IGL_region_2$length,74718404,74718404)))

features_IGL$fill <- rep(c("#101010", "#CE896A","#647C9C"), c(nrow(IGL_region_2),1,1))
features_IGL$color <- rep(c("#101010", "#CE896A","#647C9C"), c(nrow(IGL_region_2),1,1))
features_IGL$height <- c(0.04)
features_IGL$featureLayerID <-
  rep(c("bottom.IGL.allel", "median.contig", "top.chr"), c(nrow(IGL_region_2), 1, 1))
names(features_IGL) <- rep(c("IGL allel", "ctg000010", "chr17"), c(nrow(IGL_region_2), 1, 1))
x <- lolliplot(IGL_gr, features_IGL, cex = 0.6, yaxis = F, ylab = F)
```

## Figure S9B - NCBA 1.0 IGL gene loci (without gene name)

```{r}
IGL_region_2 <- as.data.frame(fread("./newData_20220115/IGL/IGL_mapResults-20220130_cleanup.csv"))
IGL_region_2$location <- IGL_region_2$start+round(0.5*IGL_region_2$length)

## 1、IGL allel graph
IGL_gr <- GRanges("IGL", IRanges(-1, width = 1))
features_IGL <- GRanges("IGL", IRanges(start = 0, end = 72700000-72000000))

features_IGL$fill <- c("#CE896A")
features_IGL$color <- c("#CE896A")
features_IGL$height <- c(0.04)

IGL_gr$SNPsideID <- "top"
features_IGL_mul <- GRanges("IGL", IRanges(start = c(0,IGL_region_2$start),
                                           end = c(72700000-72000000,IGL_region_2$start+IGL_region_2$length)))
table(IGL_region_2$functionality)
IGL_region_2$color <- NA
IGL_region_2$color[IGL_region_2$functionality %in% "F"] <- "#c93f55"
IGL_region_2$color[IGL_region_2$functionality %in% "ORF"] <- "#ffcd12"
IGL_region_2$color[IGL_region_2$functionality %in% "P"] <- "#208cc0"
IGL_region_2$color[IGL_region_2$functionality %in% "" | IGL_region_2$functionality %in% NA] <- "#828282"

IGL_allel_num <- nrow(IGL_region_2)
# features_IGL_mul$fill <- rep(c("#44435F","#CE896A"), c(IGL_allel_num,1))
# features_IGL_mul$color <- rep(c("#44435F","#CE896A"), c(IGL_allel_num,1))
features_IGL_mul$fill <- c("#44435F",IGL_region_2$color)
features_IGL_mul$color <- c("#44435F",IGL_region_2$color)
features_IGL_mul$height <- rep(0.04, length(features_IGL_mul))
# features_IGL_mul$height[4:6] <- list(unit(1/8, "inches"),
#                                  unit(0.5, "lines"),
#                                  unit(.2, "char"))
features_IGL_mul$featureLayerID <-
  rep(c("bottom.contig", "top.IGL.allel"), c(1,IGL_allel_num))
names(features_IGL_mul) <- rep(c("contig","IGL allel"),c(1,IGL_allel_num))
# x <- lolliplot(IGL_gr, features_IGL_mul, cex = 0.6, yaxis = F, ylab = F)
xaxis_tmp <- c(0,100000,200000,300000,400000,500000,600000,700000)
names(xaxis_tmp) <- c("72Mb","72.1Mb","72.2Mb","72.3Mb","72.4Mb","72.5Mb","72.6Mb","72.7Mb")
x <- lolliplot(IGL_gr, features_IGL_mul, cex = 0.6, yaxis = F, ylab = F,xaxis = xaxis_tmp)

```



## Figure S9B - NCBA 1.0 IGL region alignment (read length > 20Kb)

```{r}
options(ucscChromosomeNames=FALSE)
gTrack <- GenomeAxisTrack(cex = 1)

chr17_igl_allel_loci <- as.data.frame(fread("./newData_20220115/chrContigloci.csv"))
chr17_igl_allel_loci[17,]

color_tmp <- list(c("#749e89","#c399a2"))

## 
chrom <- "ctg000010_np12121212"
afrom <- 72000000
ato <- 72700000
alTrack <- Gviz:::.import.bam.alignments("/home/share/litt/proj/cow/results/longAlignment/contigs/ont_ctg_immuno_loci_20kb_filtered.bam", GRanges(chrom, IRanges(afrom, ato)))
alTrack <- AlignmentsTrack(sort(alTrack))
plotTracks(c(gTrack, alTrack), chromosome = "ctg000010_np12121212", from = afrom, to = ato, 
             fill.reads=color_tmp[[1]][as.numeric(sort(ranges(alTrack)$readStrand))],
             col.reads=NA, 
             alpha.reads=1)


## 
chrom <- "ctg000010_np12121212"
afrom <- 72000000
ato <- 72700000
alTrack <- Gviz:::.import.bam.alignments("/home/share/litt/proj/cow/results/longAlignment/contigs/ont_ctg_immuno_loci_IGL_72.36_72.44Mb_filtered_5k.bam", GRanges(chrom, IRanges(afrom, ato)))
alTrack <- AlignmentsTrack(sort(alTrack))
plotTracks(c(gTrack, alTrack), chromosome = "ctg000010_np12121212", from = afrom, to = ato, 
             fill.reads=color_tmp[[1]][as.numeric(sort(ranges(alTrack)$readStrand))],
             col.reads=NA, 
             alpha.reads=1)



## 
chrom <- "ctg000010_np12121212"
afrom <- 72000000
ato <- 72700000
alTrack <- Gviz:::.import.bam.alignments("/home/share/litt/proj/cow/results/longAlignment/contigs/ont_ctg_immuno_loci_20kb_filtered_merge.bam", GRanges(chrom, IRanges(afrom, ato)))
alTrack <- AlignmentsTrack(sort(alTrack))
plotTracks(c(gTrack, alTrack), chromosome = "ctg000010_np12121212", from = afrom, to = ato, 
             fill.reads=color_tmp[[1]][as.numeric(sort(ranges(alTrack)$readStrand))],
             col.reads=NA, 
             alpha.reads=1)
```

